### DISCUSSION TITLE
ADC module, supported hardware lists, GDS in production?

### DISCUSSION THREAD

**Discussion by timmyd7777** [nasa/fprime#1097] [Q&A]
We're new to F prime, and trying to decide whether to base our future flight software on F-prime vs. cFS vs. Kubos vs. something else.  I've read the overviews, tutorials, and documentation, but not started building code yet.  So these are "beginner" questions.

1. Is there a list of third-party hardware - reaction wheels, star trackers, radios, EPSs, etc. - that F-prime contains drivers or support for?  Or are we expected to roll our own, using ICDs or communication protocols supplied by the third-party hardware vendors?  Are there any open-source examples of F prime "plugins" for third-party hardware, other than the GPS module in the tutorials?
2. Is there a higher-level ADC module available anywhere for F-prime?  I mean code which implements a "de-tumble the spacecraft" algorithm, or a "hold steady nadir-pointing attitude" algorithm, that could take inputs from star trackers/gyros, etc. and send commands to reaction wheels, to stabilize a spacecraft.  Yes, this is non-trivial and mission specific.  I'm just wondering if anyone has done something like this for F-prime, and if there's source code out there which might serve as a model.
3. The F-prime GDS is intended for use in the lab, sending commands and receiving telemetry.  I'm wondering if anyone has used GDS in production, with a live spacecraft, talking through an existing ground station network like AWS, K-Sat, Atlas, etc.  Could GDS potentially be used live in production?  What are the pitfalls, etc.?

Our intended application of F-prime is for earth-orbiting spacecraft, not interplanetary.  Thanks!

**Comment by rawinza555** [comment]
Not a mod but have been working on the similar project as yours for a while now
1.) You gotta roll your own. Basically just check the device's datasheet , libraries and APIs and implement them in F'. I have managed to use F' to interact with devices like an accelerometer, nRF24L01 and SPIRIT 1 RF module. 
2.) I believe you might have to make your own port that connect sensors to reaction wheel driver and have some sort of algorithm for sensor data input handler to do stuff like that. I don't think F' has that. F' only provide you basic lego blocks for command and data handling and then we build stuff on top of it. 
3.) One of the thing that I don't like about GDS is it only show TM data in numbers. GDS doesn't do well in data visualization. In many cases, it's easier to visualize stuff in graphs or some sort of a diagram, which is really hard to modify GDS to. I think it's better to use COSMOS or OPENMCT or even create your own front end of the MCT.

**Reply by Joshua-Anderson** [reply]
I second what @rawinza555 said!

> Is there a higher-level ADC module available anywhere for F-prime?

No higher level ADC components are available within the F' framework. Higher-level, less-generic logic like ADC components are usually developed directly for specific missions.

> Could GDS potentially be used live in production? What are the pitfalls, etc.?

As you mentioned, the F' GDS is designed specifically for ground testing, with the assumption the projects will shift to a production GDS system for flight operations. That being said, there's no reason that the F' GDS can't be used for flight operations - it's just missing useful features like robust graphing support and consolidation of received telemetry into a database. If there is interest in using other GDS systems, using the fprime-gds libraries for decoding F' data, it should be fairly strait forward to integrate F' projects with other GDS systems.

**Reply by timmyd7777** [reply]
All right, thanks very much guys.  On my cursory review, f-prime seems much easier to get started with than cFS (the other big open source contender for this role).  And cFS seems to have the same drawbacks, so I don't see a clear winner yet.  One last question:

Who is the "responsible party" at JPL that maintains f-prime, and how do we contact them?  cFS lists an email address and a specific office at NASA Goddard.  I could not find an equivalent contact for f-prime.  My management would see lack of such a contact as a strike against f-prime, and a point in favor of cFS.

**Comment by timcanham** [comment]
We have a team here at JPL that maintains F Prime. We do all our interactions here on GitHub via questions like yours. If you would like to have more extensive conversations, perhaps we can set up a telecon with you and discuss how you plan to (potentially) use F Prime.
